drivers/bus/acpi/eval.c

```diff
@@ -1,255 +1,804 @@
+        }
+
+        default:
+        {
+            DPRINT1("Unsupported input buffer signature: 0x%lx\n", EvalInputBuffer->Signature);
+
+            return STATUS_INVALID_PARAMETER_1;
+        }
+    }
+
+    return STATUS_SUCCESS;
+}
+
+static
+CODE_SEG("PAGE")
+VOID
+EvalFreeParameterArgument(
+    _In_ ACPI_OBJECT* Arg,
+    _In_ ULONG Depth)
+{
+    ULONG i;
+
+    PAGED_CODE();
+
+    if (Depth >= ACPI_MAX_PACKAGE_DEPTH)
+    {
+        ASSERT(FALSE);
+        return;
+    }
+
+    if (Arg->Type == ACPI_TYPE_PACKAGE)
+    {
+        for (i = 0; i < Arg->Package.Count; i++)
+        {
+            EvalFreeParameterArgument(&Arg->Package.Elements[i], Depth + 1);
+        }
+
+        /* Check if the package isn't empty, and free it */
+        if (Arg->Package.Elements)
+            ExFreePoolWithTag(Arg->Package.Elements, TAG_ACPI_PACKAGE_LIST);
+    }
+}
+
+static
+CODE_SEG("PAGE")
+VOID
+EvalFreeParametersList(
+    _In_ ACPI_OBJECT_LIST* ParamList)
+{
+    ACPI_OBJECT* Arg;
+    ULONG i;
+
+    PAGED_CODE();
+
+    Arg = ParamList->Pointer;
+    for (i = 0; i < ParamList->Count; i++)
+    {
+        EvalFreeParameterArgument(Arg++, 0);
+    }
+
+    ExFreePoolWithTag(ParamList->Pointer, TAG_ACPI_PARAMETERS_LIST);
+}
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+EvalAcpiStatusToNtStatus(
+    _In_ ACPI_STATUS AcpiStatus)
+{
+    PAGED_CODE();
+
+    if (ACPI_ENV_EXCEPTION(AcpiStatus))
+    {
+        switch (AcpiStatus)
+        {
+            case AE_NOT_FOUND:
+            case AE_NOT_EXIST:
+                return STATUS_OBJECT_NAME_NOT_FOUND;
+
+            case AE_NO_MEMORY:
+                return STATUS_INSUFFICIENT_RESOURCES;
+
+            case AE_SUPPORT:
+                return STATUS_ACPI_INCORRECT_ARGUMENT_COUNT;
+
+            case AE_TIME:
+                return STATUS_IO_TIMEOUT;
+
+            case AE_NO_HARDWARE_RESPONSE:
+                return STATUS_IO_DEVICE_ERROR;
+
+            case AE_STACK_OVERFLOW:
+                return STATUS_ACPI_STACK_OVERFLOW;
+
+            default:
                 break;
+        }
+    }
+
+    if (ACPI_AML_EXCEPTION(AcpiStatus))
+    {
+        switch (AcpiStatus)
+        {
+            case AE_AML_UNINITIALIZED_ARG:
+                return STATUS_ACPI_INCORRECT_ARGUMENT_COUNT;
 
             default:
-                DPRINT1("Unsupported element type %lu\n", Package->Package.Elements[i].Type);
-                return STATUS_UNSUCCESSFUL;
+                break;
         }
+    }
 
-        Ptr = ACPI_METHOD_NEXT_ARGUMENT(Ptr);
+    return STATUS_UNSUCCESSFUL;
+}
+
+static
+CODE_SEG("PAGE")
+ACPI_STATUS
+EvalEvaluateObject(
+    _In_ PPDO_DEVICE_DATA DeviceData,
+    _In_ PACPI_EVAL_INPUT_BUFFER EvalInputBuffer,
+    _In_ ACPI_OBJECT_LIST* ParamList,
+    _In_ ACPI_BUFFER* ReturnBuffer)
+{
+    CHAR MethodName[4 + 1];
+
+    PAGED_CODE();
+
+    RtlCopyMemory(MethodName,
+                  EvalInputBuffer->MethodName,
+                  sizeof(EvalInputBuffer->MethodName));
```

Comment:
Alright so technically EvalInputBuffer->MethodName is never going to be above the size of MethodName but let's make this clear in code.

1) make a Preprocessor define to signify the maximum MethodName size and use that.  (That's a nit)
2) this size parameter should be the value of the preprocessor define mentioned in 1 instead of the size of the data we're reading.

you're using a hard coded number for the dest buffer, so the size of information we're copying into dest should be hard coded as well.

Reply:
I even wonder why the local MethodName variable couldn't be defined as follows:
```c
    CHAR MethodName[sizeof(EvalInputBuffer->MethodName)];
```
Otherwise, this "4 + 1" looks like what the `ACPI_OBJECT_NAME_LENGTH` define is about.

Reply:
Reused the existing `ACPI_OBJECT_NAME_LENGTH` define.